scipy/spatial/src/distance_metrics.h

```diff
@@ -169,6 +169,250 @@ void transform_reduce_2d_(
+        }
+    }
+    for (; i < x.shape[0]; ++i) {
+        const T* x_row = &x(i, 0);
+        AccumulateType dist = {};
+        for (intptr_t j = 0; j < x.shape[1]; ++j) {
+            auto val = map(x_row[j * xs]);
+            dist = reduce(dist, val);
+        }
+        out[i] = project(dist);
+    }
+}
+
+template <int ilp_factor=2, typename T,
+          typename TransformFunc,
+          typename ProjectFunc = Identity,
+          typename ReduceFunc = Plus>
+void transform_reduce_2d_(
+    T* out, StridedView2D<T> x, StridedView2D<const T> w,
+    const TransformFunc& map,
+    const ProjectFunc& project = Identity{},
+    const ReduceFunc& reduce = Plus{}) {
+    intptr_t i = 0;
+    intptr_t xs = x.strides[1], ws = w.strides[1];
+    // Result type of calling map
+    using AccumulateType = typename std::decay<decltype(
+        map(std::declval<T>(), std::declval<T>()))>::type;
+
+    for (; i + (ilp_factor - 1) < x.shape[0]; i += ilp_factor) {
+        const T* x_rows[ilp_factor];
+        const T* w_rows[ilp_factor];
+        ForceUnroll<ilp_factor>{}([&](int k) {
+            x_rows[k] = &x(i + k, 0);
+            w_rows[k] = &w(i + k, 0);
+        });
+
+        AccumulateType dist[ilp_factor] = {};
+        for (intptr_t j = 0; j < x.shape[1]; ++j) {
+            ForceUnroll<ilp_factor>{}([&](int k) {
+                auto val = map(x_rows[k][j * xs], w_rows[k][j * ws]);
+                dist[k] = reduce(dist[k], val);
+            });
+        }
+
+        ForceUnroll<ilp_factor>{}([&](int k) {
+            out[i + k] = project(dist[k]);
+        });
+    }
+    for (; i < x.shape[0]; ++i) {
+        const T* x_row = &x(i, 0);
+        const T* w_row = &w(i, 0);
+        AccumulateType dist = {};
+        for (intptr_t j = 0; j < x.shape[1]; ++j) {
+            auto val = map(x_row[j * xs], w_row[j * ws]);
+            dist = reduce(dist, val);
+        }
+        out[i] = project(dist);
+    }
+}
+
+template <int ilp_factor=4,
+          typename T,
+          typename TransformFunc>
+void transform_reduce_2d_(
+    StridedView2D<T> x, T* rowscales,
+    const TransformFunc& map) {
+    intptr_t xs = x.strides[1];
+
+    intptr_t i = 0;
+    if (xs == 1) {
+        for (; i + (ilp_factor - 1) < x.shape[0]; i += ilp_factor) {
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                ForceUnroll<ilp_factor>{}([&](int k) {
+                    x(i + k, j) = map(x(i + k, j), rowscales[i + k]);
+                });
+            }
+        }
+    } else {
+        for (; i + (ilp_factor - 1) < x.shape[0]; i += ilp_factor) {
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                ForceUnroll<ilp_factor>{}([&](int k) {
+                    x(i + k, j) = map(x(i + k, j), rowscales[i + k]);
+                });
+            }
+        }
+    }
+    for (; i < x.shape[0]; ++i) {
+        for (intptr_t j = 0; j < x.shape[1]; ++j) {
+            x(i, j) = map(x(i, j), rowscales[i]);
+        }
+    }
+}
```

Comment:
The three new `transform_reduce_2d_` only preprocess inputs (weighted and unweighted). The last one doesn't have a reduce step because it just needs updating of the input by division, subtraction, etc. The first two do have map-reduce step but only accept `x` (no `y`).